### ISSUE TITLE
False positive of "Dead instanceof" hint

### ISSUE THREAD

**Issue by markiewb** [markiewb/nb-additional-hints#50]
ACTUAL:

```
java.util.Set a = null; if (a instanceof java.util.HashSet) {} // ERROR
java.util.Set e = null; if (e instanceof java.util.ArrayList) {} //ERROR
```

EXPECTED:
```
java.util.Set a = null; if (a instanceof java.util.HashSet) {} // *NO* ERROR
java.util.Set e = null; if (e instanceof java.util.ArrayList) {} //ERROR
```

**Comment by FrantaM** [comment]
Why is `(e instanceof java.util.ArrayList)` an error? While `ArrayList` itself cannot be cast to `Set`, you can have a _subclass_ of `ArrayList` that can. See http://ideone.com/sqQbM0.

**Comment by markiewb** [comment]
The hint only checks the compile-type of "e". "e" is an Set, so internally
it checks for (java.util.Set instanceof java.util.ArrayList), which returns
false.



With kind regards, markiewb

2015-04-16 11:52 GMT+02:00 FrantaM <notifications@github.com>:

**Comment by FrantaM** [comment]
I can see _what_ it does I'm asking as to _why_. Instance of `ArrayList` can be an instanceof `Set` at the same time so this is still a false positive.

**Comment by markiewb** [comment]
My usecase is to catch the following construct

```
            java.util.Set set = xxx;
            if (set instanceof java.util.ArrayList) {
                java.util.ArrayList list = set; //<- compile error
            }
```

With kind regards, markiewb

2015-04-16 12:39 GMT+02:00 FrantaM <notifications@github.com>:

**Comment by btbouwens** [comment]
Indeed this bug is still active: I see this for `f instanceof Inter` for an interface `Inter` which isn't implemented by the type of `f` itself, but by possible subclasses.

**Comment by jrb0001** [comment]
Still not fixed in v1.6.0.3-SNAPSHOT:
```java
InterfaceA something = new ClassBExtendsCImplementsA();
System.out.println(something instanceof ClassC); // Runtime: true, Netbeans: "Dead instanceof. InterfaceA cannot be assigned to ClassC"
```

**Comment by markiewb** [comment]
@btbouwens, @jrb0001 , @FrantaM : You are all correct. That hint makes no sense. I will remove it via #70 

NB already warns you about invalid combinations.
```
// For example
if (new Double(42) instanceof java.util.Map) {}
```
